gh-135552: Add tests that check if weakref for tp_subclasses cleared after finalization

sergey-miryanov opened the pull request:
Tests for https://github.com/python/cpython/pull/136189

<!-- gh-issue-number: gh-135552 -->
* Issue: gh-135552
<!-- /gh-issue-number -->

fxeqxmulfx committed:
Add tests

sergey-miryanov committed:
Organize tests

sergey-miryanov committed:
Simplify tests

sergey-miryanov committed:
Remove extra line

sergey-miryanov committed:
Add extra comments

sergey-miryanov committed:
Add tests for checking that weakrefs clears before and after finalization

neonene commented on Lib/test/test_gc.py:
```diff
@@ -1125,6 +1125,82 @@ def test_something(self):
+        """
+        _, _, stderr = assert_python_ok("-c", code)
+        assert b"Instance weakref is not None" not in stderr
+        assert b"Type weakref is not Class" not in stderr
```
I do not know the issue, but the "not in" assertions look unreliable.  The test will pass forever once the strings are accidentally changed.  Also, any reason not to use `self.assert*()`?

sergey-miryanov replied:
Thanks! I believe it is better now. WDYT?

neonene replied:
Again, why do the `assert` statements here make sense over consistency in the code base?  If you have complaints about `self.assert*()`, you need to get consensus in dpo first.  Other than that the assertions look fine.

sergey-miryanov replied:
Ough! Sorry! I misunderstood you about asserts. Thanks for patience and pointing me out!
Fixed.

sergey-miryanov committed:
Make tests a little bit reliable

sergey-miryanov committed:
Simplify tests

sergey-miryanov committed:
Use unittest asserts

sergey-miryanov committed:
Move test_clearing_weakrefs_in_gc to test_weakref

kumaraditya303 commented on Lib/test/test_gc.py:
```diff
@@ -1125,6 +1125,37 @@ def test_something(self):
         assert_python_ok("-c", source)
 
+    def test_do_not_cleanup_type_subclasses_before_finalization(self):
+        # https://github.com/python/cpython/issues/135552
```
```suggestion
        #  See https://github.com/python/cpython/issues/135552
```

kumaraditya303 commented on Lib/test/test_gc.py:
```diff
@@ -1125,6 +1125,37 @@ def test_something(self):
+        # If we cleanup weakrefs for tp_subclasses before calling
+        # the finalizer (__del__) then the line `fail = BaseNode.next.next`
+        # should fail because we are trying to access a subclass
+        # atribute. But subclass type cache was not properly invalidated.
```
```suggestion
        # attribute. But subclass type cache was not properly invalidated.
```

sergey-miryanov committed:
Apply suggestions from code review

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

nascheme committed:
Merge branch 'main' into gh-135552-add-tests-gc-weakref-tp_subclasses

nascheme committed:
Revise test_clearing_weakrefs_in_gc test

Types are not treated specially, it only matters if the weakref has a callback or not.
